lib/local-cluster/lib/Control/Monitoring/Folder.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Control.Monitoring.Folder
+    ( mkFoldingMonitor
+    )
+where
+
+import Prelude
+
+import Control.Comonad
+    ( Comonad (duplicate, extract)
+    , ($>)
+    )
+import Control.Foldl
+    ( Fold
+    )
+import Control.Monitoring.Monitor
+    ( MonitorState (..)
+    , PullingOrNotPulling (..)
+    )
+import Control.Monitoring.Tracing
+    ( NotPulling (..)
+    , Pulling (..)
+    , Tracing (..)
+    )
+import UnliftIO
+    ( MonadIO
+    , atomically
+    , liftIO
+    , modifyTVar
+    , newEmptyTMVarIO
+    , newTVarIO
+    , putTMVar
+    , readTVarIO
+    , takeTMVar
+    )
+
+import qualified Control.Foldl as F
+
+-- | Create a NotPulling tracer from a `Fold`
+mkFoldingMonitor
+    :: MonadIO m
+    => m ctx
+    -- ^ How to get the context for each trace
+    -> Fold (ctx, a) b
+    -- ^ How to fold the context and the value
+    -> MonitorState
+    -> m (PullingOrNotPulling m a b)
```

Comment:
I have a feeling that the concerns are not captured well, yet — essentially, `PullingOrNotPulling` is an automaton / Mealy machine that intertwines the monad `m` with the folding functionality, but it has a less than ideal name. Actually I think that the pure core is a `Scan` rather than a `Fold` (— for each new input `a`, you get a new trace value `b`).

But anyway, I think the separation of concerns is good enough — we will typically call it in a combination `mkMonitor <=< mkFoldingMonitor` and there it's a good implementation that separates `TVar` well enough (one mutable state for keeping track of whether we are pulling or not pulling; another mutable state for the observed trace.)

Reply:
It's more like a moore machine. In our case we have at least to persist the readiness status, but in general we will be able to permanently answer to "which phases have been started?". So the folding core of moore machines. But this is one implementation only. As for the name I will have another try.